File: python/src/rivu_server_sdk/resume.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Literal

from .event_store import Envelope, InMemoryRingBufferEventStore
from .seq import SeqAllocator
from .snapshot_store import SqliteSnapshotStore


ResumeKind = Literal["replay", "snapshot", "not_found"]


@dataclass(frozen=True)
class ResumeResult:
    kind: ResumeKind
    envelopes: list[Envelope]


def resume_replay(
    *,
    thread_id: str,
    resume_from: int,
    event_store: InMemoryRingBufferEventStore,
    snapshot_store: SqliteSnapshotStore,
    seq_allocator: SeqAllocator | None = None,
) -> ResumeResult:
    """Return envelopes to send for `resumeFrom`.

    - Prefer contiguous replay from the event store.
    - If replay is incomplete, fall back to a `STATE_SNAPSHOT` at `seq = resume_from + 1`.
    """
    replay = event_store.replay_after(thread_id, after_seq=resume_from)
    if replay.complete:
        return ResumeResult(kind="replay", envelopes=replay.envelopes)

    snap = snapshot_store.get_latest(thread_id=thread_id)
    if snap is None:
        return ResumeResult(kind="not_found", envelopes=[])

    seq = resume_from + 1
    if seq_allocator is not None:
        seq_allocator.observe(thread_id, seq)

    snapshot_event: dict[str, Any] = {"type": "STATE_SNAPSHOT", "snapshot": snap.shared_state}
    return ResumeResult(kind="snapshot", envelopes=[{"seq": seq, "event": snapshot_event}])

File: crates/rivu-server-sdk/src/resume.rs
use serde_json::{json, Value};

use crate::event_store::{Envelope, InMemoryRingBufferEventStore};
use crate::seq::SeqAllocator;
use crate::snapshot_store::SqliteSnapshotStore;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ResumeKind {
    Replay,
    Snapshot,
    NotFound,
}

#[derive(Debug, Clone)]
pub struct ResumeResult {
    pub kind: ResumeKind,
    pub envelopes: Vec<Envelope>,
}

#[derive(thiserror::Error, Debug)]
pub enum ResumeError {
    #[error(transparent)]
    EventStore(#[from] crate::event_store::EventStoreError),
    #[error(transparent)]
    SnapshotStore(#[from] crate::snapshot_store::SnapshotStoreError),
    #[error(transparent)]
    Seq(#[from] crate::seq::SeqError),
}

pub fn resume_replay(
    thread_id: &str,
    resume_from: u64,
    event_store: &InMemoryRingBufferEventStore,
    snapshot_store: &SqliteSnapshotStore,
    seq_allocator: Option<&SeqAllocator>,
) -> Result<ResumeResult, ResumeError> {
    let replay = event_store.replay_after(thread_id, resume_from)?;
    if replay.complete {
        return Ok(ResumeResult {
            kind: ResumeKind::Replay,
            envelopes: replay.envelopes,
        });
    }

    let snapshot = snapshot_store.get_latest(thread_id)?;
    let Some(snap) = snapshot else {
        return Ok(ResumeResult {
            kind: ResumeKind::NotFound,
            envelopes: vec![],
        });
    };

    let seq = resume_from + 1;
    if let Some(alloc) = seq_allocator {
        alloc.observe(thread_id, seq)?;
    }

    let event: Value = json!({ "type": "STATE_SNAPSHOT", "snapshot": snap.shared_state });
    Ok(ResumeResult {
        kind: ResumeKind::Snapshot,
        envelopes: vec![Envelope { seq, event }],
    })
}

